backend/server.js

```diff
@@ -17,10 +17,25 @@ import {
   createPushErrorResponse,
   createServer
 } from './pubsub.js'
-import { pushServerActionhandlers } from './push.js'
-// $FlowFixMe[cannot-resolve-module]
-import { webcrypto } from 'node:crypto'
+import { addChannelToSubscription, deleteChannelFromSubscription, pushServerActionhandlers, subscriptionInfoWrapper } from './push.js'
 import { GIMessage } from '~/shared/domains/chelonia/GIMessage.js'
+import type { SubMessage, UnsubMessage } from '~/shared/pubsub.js'
+
+// Node.js version 18 and lower don't have global.crypto defined
+// by default
+if (
+  !('crypto' in global) &&
+  typeof require === 'function'
+) {
+  const { webcrypto } = require('crypto')
+  if (webcrypto) {
+    Object.defineProperty(global, 'crypto', {
+      'enumerable': true,
+      'configurable': true,
+      'get': () => webcrypto
+    })
+  }
+}
```

Comment:
I was looking at the rest of the file and noticed this:

```js
    // If the contract has been removed or the height hasn't been updated,
    // there's nothing to persist
    if (!cheloniaState.contracts[contractID] || cheloniaState.contracts[contractID].height < deserializedHEAD.head.height) {
      return
    }
```

I'm a little confused by that second condition.... is it backwards? (Should `<` be `>`?)

Not quite understanding, can you explain?

Reply:
Good that you asked, it's good to review these from time to time. I believe this to be correct. The intent is that only processed events should be persisted (i.e., if processing errored, the state shouldn't be persisted).

With this in mind, we have the first check:

1. `cheloniaState.contracts[contractID]`: This means the `cheloniaState.contracts` entry doesn't exist. This could be because the contract has been removed, or because it was an  OP_CONTRACT` that couldn't be processed. Either way, as the comment says, there's nothing to persist.
2. `cheloniaState.contracts[contractID].height < deserializedHEAD.head.height`: When a message gets correctly processed, we expect that `cheloniaState.contracts[contractID].height === deserializedHEAD.head.height`. If there were other events processed since, we also expect that the state's height will be higher than the current message (i.e., `cheloniaState.contracts[contractID].height > deserializedHEAD.head.height`). The condition being checked however is for a third possibility, which is that the current message was _not_ processed, hence the height was not updated. In this case, there is also nothing to persist. (There is a fourth case, which isn't handled: none of `<`, `>` or `===` are true, i.e., if either or both of the values are `NaN`. This shouldn't be possible.)